src/core/lib/security/security_connector/tls/tls_security_connector.cc

```diff
@@ -62,34 +62,41 @@ tsi_ssl_pem_key_cert_pair* ConvertToTsiPemKeyCertPair(
 
 }  // namespace
 
-/** -- Util function to fetch TLS server/channel credentials. -- */
 grpc_status_code TlsFetchKeyMaterials(
     const grpc_core::RefCountedPtr<grpc_tls_key_materials_config>&
         key_materials_config,
```

Comment:
May I know what are the benefits we can provide when defining this field as "a const reference to a RefCountedPtr"?
It gives me a sense that `key_materials_config` is somewhat similar with `reload_status`, in a way that they are both provided by the caller, and populated inside this function.
Since reload_status is a plain pointer, can we just make key_materials_config a RefCountedPtr?

Reply:
Sorry maybe I misunderstood - is the question at the end supposed to be: can we just make `key_materials_config` a raw pointer?

If so, then yes I think we can - none of the logic inside the `TlsFetchKeyMaterials` method would change. Moreover, all of the callers of `TlsFetchKeyMaterials` already pass in a raw pointer.

This method was written by @yihuazhang so perhaps we should ask him as well. Yihua: is there any reason to keep the `RefCountedPtr` here?

Reply:
Sorry, I probably should rephrase my last sentence to 
"Since reload_status is a plain pointer, can we make key_materials_config a RefCountedPtr, or even a plain pointer?"
I think by using a RefCountedPtr, we could imply that `TlsFetchKeyMaterials` shares the ownership of `key_materials_config`, which looks correct because we need to modify this parameter in `TlsFetchKeyMaterials`. So I think it's better using RefCountedPtr than a plain pointer.
What I don't understand is why we need a const reference...I might be over picky about this, but I really want to get an idea of the philosophy when designing these APIs...

Reply:
I found this post explains this pretty well: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/
To my understanding, we use something like "const shared_ptr &" when we want to imply the sharing semantics, but don't want the pointer to be copied. I think I understand it now...
Sorry for the noise here, @matthewstevenson88 ! I will resolve this right away :-)